src/ray/common/ray_syncer/ray_syncer.h

```diff
@@ -0,0 +1,212 @@
+///       and make sure not sending the information the remote node knows.
+///    2. NodeState: keeps track of the local status, similar to NodeSyncConnection,
+//        but it's for local node.
+class RaySyncer {
+ public:
+  /// Constructor of RaySyncer
+  ///
+  /// \param io_context The io context for this component.
+  /// \param node_id The id of current node.
+  RaySyncer(instrumented_io_context &io_context, const std::string &node_id);
+  ~RaySyncer();
+
+  /// Connect to a node.
+  /// TODO (iycheng): Introduce grpc channel pool and use node_id
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::unique_ptr<NodeSyncConnection> connection);
+
+  /// Connect to a node.
+  /// TODO (iycheng): Introduce grpc channel pool and use node_id
+  /// for the connection.
+  ///
+  /// \param connection The connection to the remote node.
+  void Connect(std::shared_ptr<grpc::Channel> channel);
+
+  void Disconnect(const std::string &node_id);
+
+  /// Register the components to the syncer module. Syncer will make sure eventually
+  /// it'll have a global view of the cluster.
+  ///
+  /// \param component_id The component to sync.
+  /// \param reporter The local component to be broadcasted.
+  /// \param receiver The snapshot of the component in the cluster.
+  /// \param pull_from_reporter_interval_ms The frequence to pull a message from reporter
+  /// and push it to sending queue.
+  bool Register(RayComponentId component_id,
+                const ReporterInterface *reporter,
+                ReceiverInterface *receiver,
+                int64_t pull_from_reporter_interval_ms = 100);
+
+  /// Function to broadcast the messages to other nodes.
+  /// A message will be sent to a node if that node doesn't have this message.
+  /// The message can be generated by local reporter or received by the other node.
+  ///
+  /// \param message The message to be broadcasted.
+  void BroadcastMessage(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Get the current node id.
+  const std::string &GetNodeId() const { return node_id_; }
+
+  /// Get the io_context used by RaySyncer.
+  instrumented_io_context &GetIOContext() { return io_context_; }
+
+  NodeSyncConnection *GetSyncConnection(const std::string &node_id) {
+    auto iter = sync_connections_.find(node_id);
+    if (iter == sync_connections_.end()) {
+      return nullptr;
+    }
+    return iter->second.get();
+  }
+
+ private:
+  /// io_context for this thread
+  instrumented_io_context &io_context_;
+
+  /// The current node id.
+  const std::string node_id_;
+
+  /// Manage connections
+  absl::flat_hash_map<std::string, std::unique_ptr<NodeSyncConnection>> sync_connections_;
+
+  /// The local node state
+  std::unique_ptr<NodeState> node_state_;
+
+  /// If the field is set to be true, it means, message generated or received
+  /// is needed to be sent to other nodes.
+  Array<bool> component_broadcast_;
```

Comment:
Do we need this now? (do we have any pattern that doesn't require the broadcast now?)

Reply:
we need. resource loads doesn't need to be broadcasted.